### PR TITLE
[1/N][data, omni, tests] feat: add Omni-Preference offline MLLM DPO dataset pipeline

### PR DESCRIPTION
**Pull request by wtomin**
## Summary

Adressing RFC  #266. Part of #269.

Add end-to-end data infrastructure for **offline MLLM DPO** training on the [Omni-Preference](https://huggingface.co/datasets/Omni-RRM/Omni-Preference) dataset (image / video / audio preference pairs):

- **Preprocessing**: convert `final_rl_data.jsonl` into multimodal parquet (`prompt` = media + question, `chosen` / `rejected` = candidate answers), with media-file-granularity train/test split.
- **Dataset layer**: `OfflineMLLMDPODataset`, `offline_mllm_dpo_collate_fn`, and `ModalityGroupedBatchSampler` for same-modality batching across mixed parquet sources.
- **Transform**: standalone `qwen3_omni_transform` (VeOmni-free) mirroring VeOmni's Qwen3-Omni sample processing for chosen/rejected branches.
- **Smoke data**: dummy parquet generator for e2e smoke tests.
- **Sampler hook**: extend `create_rl_sampler` in `rl_dataset.py` to load custom sampler classes from data config.
- **Docs**: dataset download, preprocessing, and parquet schema guide under `examples/dpo_trainer/data_process/` (not tied to a specific model example directory).

## Changes

| Area | File | Description |
|------|------|-------------|
| Preprocess | `examples/dpo_trainer/data_process/omni_preference_dpo_multisource.py` | JSONL → parquet for image/video/audio; media-name split; path resolution |
| Dataset | `verl_omni/utils/dataset/offline_mllm_dpo_dataset.py` | `OfflineMLLMDPODataset`, `offline_mllm_dpo_collate_fn`, `ModalityGroupedBatchSampler` |
| Transform | `verl_omni/utils/dataset/qwen3_omni_transform.py` | Qwen3-Omni multimodal sample transform (VeOmni-free) |
| Sampler | `verl_omni/utils/dataset/rl_dataset.py` | `create_rl_sampler` supports `data.sampler.class_path` / `class_name` |
| E2E data | `tests/special_e2e/create_dummy_omni_preference_dpo_data.py` | Tiny image/video/audio parquet + media for smoke tests |
| CPU tests | `tests/utils/test_offline_mllm_dpo_dataset_on_cpu.py` | 17 CPU unit tests for dataset helpers, collate, sampler, and preprocessing |
| Docs | `examples/dpo_trainer/data_process/omni_preference_dpo_dataset.md` | Download, preprocess, parquet schema, troubleshooting |

## Test plan

- [x] Generate dummy parquet:
  ```bash
  python tests/special_e2e/create_dummy_omni_preference_dpo_data.py \
     --local_save_dir /tmp/omni_pref_dummy

- [x] CPU unit tests:
  ```bash
  pytest tests/utils/test_offline_mllm_dpo_dataset_on_cpu.py -v

- [x] (Optional, with full Omni-Preference clone) Run preprocessing:

python examples/dpo_trainer/data_process/omni_preference_dpo_multisource.py \
  --dataset_root $HOME/Omni-Preference \
  --output_dir $HOME/Omni-Preference/parquet_dpo \
  --modalities image video audio \
  --test_ratio 0.10 \
  --seed 42

### CODE REVIEW

**Comment by zhtmike** [comment]
looks good

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+class ModalityGroupedBatchSampler(Sampler[int]):
+    """Yield indices in same-modality chunks for regular DataLoader batching.
+
+    ``StatefulDataLoader`` is configured with ``sampler=`` and ``batch_size=``,
+    not ``batch_sampler=``. This sampler therefore yields individual indices,
+    but orders them as contiguous same-modality chunks of ``batch_size``. When
+    Each chunk first samples a modality uniformly by default, or by
+    ``modality_sample_weights`` when provided, then samples rows from that
+    modality with replacement.
```
Comment (zhtmike) [inline verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:490]: instead of this, using the standard format

```
what is class is doing

Args:
   xxx

Returns:
   xxx
```

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+    def __iter__(self):
+        for batch in self._build_batches():
+            yield from batch
+
+    def __len__(self) -> int:
+        return self._length
+
+
+def offline_mllm_dpo_collate_fn(features):
```
Comment (zhtmike) [inline verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:604]: docstring for public APIs

File: verl_omni/utils/dataset/qwen3_omni_transform.py
Comment (zhtmike) [inline verl_omni/utils/dataset/qwen3_omni_transform.py:1]: should we put it model-specific transforms under pipelines/ ?

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py (hunk 1 above)
Comment (wtomin) [reply verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:490]: Fixed the docstring.

File: verl_omni/utils/dataset/qwen3_omni_transform.py
Comment (wtomin) [reply verl_omni/utils/dataset/qwen3_omni_transform.py:1]: Since `process_qwen3_omni_sample` is directly consumed by `OfflineMLLMDPODataset`, not by qwen3-omni pipeline, I suggest to keep it under `verl_omni/utils/dataset/`

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py (hunk 2 above)
Comment (wtomin) [reply verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:604]: Fixed the docstring.

File: verl_omni/utils/dataset/qwen3_omni_transform.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+        else:
+            h_bar = math.ceil(height * beta)
+            w_bar = math.ceil(width * beta)
+    return image.resize((w_bar, h_bar))
+
+
+def _fetch_images(images: Sequence[str], **kwargs) -> list[Image.Image]:
+    max_image_nums = kwargs.get("max_image_nums", len(images))
+    loaded = [Image.open(path).convert("RGB") for path in images[:max_image_nums]]
+    return [_smart_resize_image(image, **kwargs) for image in loaded]
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/utils/dataset/qwen3_omni_transform.py:101]: pls fix this

File: tests/special_e2e/create_dummy_omni_preference_dpo_data.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+from __future__ import annotations
+
+import argparse
+import os
+import struct
+import wave
+
+import av
+import pandas as pd
+from PIL import Image
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply tests/special_e2e/create_dummy_omni_preference_dpo_data.py:35]: pls udpate doc and requirments

**Comment by zhtmike** [comment]
pls check ai review comments

**Comment by wtomin** [comment]
@zhtmike Sorry I need to take some more time to debug the training convergence of #269, which might edit the dataset and batch transform func. I will synchronize the edits afterwards, so please do not merge this PR.

**Comment by wtomin** [comment]
@zhtmike #269 is making progress in training convergence, and I have synced the datasets related code to this PR.

This PR is ready to be both reviewed and merged.

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+    if isinstance(extra_info, dict) and extra_info.get("modality"):
+        return _normalise_modality(extra_info["modality"], default)
+    for key, modality in (("images", "image"), ("videos", "video"), ("audios", "audio")):
+        if _normalise_media_list(row.get(key)):
+            return modality
+    return _normalise_modality(row.get(source_name_key) or row.get("source_name"), default)
+
+
+class OfflineMLLMDPODataset(Dataset):
+    """Dataset for Omni-Preference rows consumed by Qwen3-Omni offline DPO."""
```
Comment (zhtmike) [inline verl_omni/utils/dataset/offline_mllm_dpo_dataset.py]: make a standard doctoring for API

```
What is this class doing

Args:
    xxx

Returns:
   xxxx
```

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py (hunk 1 above)
Comment (zhtmike) [reply verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:490]: Seems it is accidentally reverted

File: verl_omni/utils/dataset/offline_mllm_dpo_dataset.py [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+from typing import Any
+
+import numpy as np
+import pandas as pd
+import torch
+from omegaconf import DictConfig, OmegaConf
+from torch.utils.data import Dataset, Sampler
+
+from verl_omni.utils.dataset.qwen3_omni_transform import process_qwen3_omni_sample
+
```
Comment (zhtmike) [inline verl_omni/utils/dataset/offline_mllm_dpo_dataset.py:35]: add a `__all__ = [xxx]` for the major APIs at the top of the file

File: verl_omni/utils/dataset/qwen3_omni_transform.py
Comment (zhtmike) [reply verl_omni/utils/dataset/qwen3_omni_transform.py:1]: I see

**Comment by zhtmike** [comment]
> @zhtmike #269 is making progress in training convergence, and I have synced the datasets related code to this PR.
> 
> This PR is ready to be both reviewed and merged.

some docstring issue. make API user friendly

**Comment by wtomin** [comment]
> > @zhtmike #269 is making progress in training convergence, and I have synced the datasets related code to this PR.
> > This PR is ready to be both reviewed and merged.
> 
> some docstring issue. make API user friendly

Docstrings were fixed. PTAL.

**Review by zhtmike — APPROVED** [review]
